gflownet/envs/cube.py

```diff
@@ -0,0 +1,1317 @@
+        states_from: List,
+    ) -> TensorType["batch_size"]:
+        """
+        Computes log probabilities of forward actions.
+        """
+        # Initialize variables
+        n_states = policy_outputs.shape[0]
+        states_from_tensor = tfloat(
+            states_from, float_type=self.float, device=self.device
+        )
+        is_eos = torch.zeros(n_states, dtype=torch.bool, device=self.device)
+        logprobs_eos = torch.zeros(n_states, dtype=self.float, device=self.device)
+        logprobs_increments_rel = torch.zeros(
+            (n_states, self.n_dim), dtype=self.float, device=self.device
+        )
+        jacobian_diag = torch.ones(
+            (n_states, self.n_dim), device=self.device, dtype=self.float
+        )
+        eos_tensor = tfloat(self.eos, float_type=self.float, device=self.device)
+        # Determine source states
+        is_source = ~mask[:, 1]
+        # EOS is the only possible action if continuous actions are invalid (mask[0] is
+        # True)
+        is_eos_forced = mask[:, 0]
+        is_eos[is_eos_forced] = True
+        # Ensure that is_eos_forced does not include any source state
+        assert not torch.any(torch.logical_and(is_source, is_eos_forced))
+        # Get sampled EOS actions and get log probs from Bernoulli distribution
+        do_eos = torch.logical_and(~is_source, ~is_eos_forced)
+        if torch.any(do_eos):
+            is_eos_sampled = torch.zeros_like(do_eos)
+            is_eos_sampled[do_eos] = torch.all(actions[do_eos] == eos_tensor, dim=1)
+            is_eos[is_eos_sampled] = True
+            logits_eos = self._get_policy_eos_logit(policy_outputs)[do_eos]
+            distr_eos = Bernoulli(logits=logits_eos)
+            logprobs_eos[do_eos] = distr_eos.log_prob(
+                is_eos_sampled[do_eos].to(self.float)
+            )
+        # Get log probs of relative increments if EOS was not the sampled or forced
+        # action
+        do_increments = ~is_eos
+        if torch.any(do_increments):
+            # Get absolute increments
+            increments_abs = actions[do_increments]
+            # Get minimum increments
+            min_increments = torch.full_like(
+                increments_abs, self.min_incr, dtype=self.float, device=self.device
+            )
+            min_increments[is_source[do_increments]] = 0.0
+            # Get relative increments
+            increments_rel = self.absolute_to_relative_increments(
+                states_from_tensor[do_increments],
+                increments_abs,
+                min_increments,
+                self.max_val,
+                is_backward=False,
+            )
+            # Get logprobs
+            distr_increments = self._make_increments_distribution(
+                policy_outputs[do_increments]
+            )
+            # Clamp because increments of 0.0 or 1.0 would yield nan
+            logprobs_increments_rel[do_increments] = distr_increments.log_prob(
```

Comment:
As discussed on Slack, `get_logprobs` (both backward and forward) doesn't really work properly, at the very least with correlation estimation from dataset: in particular, in the highlighted line, `clamp` will mask the fact that some of the `increments_rel` are in fact `inf` (which I assume was not the intention).

Reply:
There may be something to fix regarding the source state, but regarding infinite increments, they should never make it to this line since only finite increments will be included in the filter `do_increments`. In fact, if we passed infinte increments to the Beta distribution it would complain badly.

Reply:
Infs are being replaced by numbers due to clamping. But try inserting
```
if torch.any(torch.isinf(increments_rel)):
    import pdb; pdb.set_trace()
```
in the above line (just before clamping) and try running `+experiments=ccube/ccube_owl proxy=corners` (the same should be happening with backward).

Reply:
Oh oops, OK, I'll look into this.

Reply:
Ok, this is related to the source states. The environment does not allow EOS from the source, but the test set contains a (terminating) state which is the source... So weird things happen. I guess I will finally have to face this issue seriously.

Reply:
Did you end up addressing this?

Reply:
Yes, it was a bunch of changes that I made after this so hard to summarise exactly what it was. But I've just tried again with the breakpoint and it does not stop.